ArduCopter/events.cpp

```diff
@@ -4,152 +4,235 @@
  *       This event will be called when the failsafe changes
  *       boolean failsafe reflects the current state
  */
+
+bool Copter::failsafe_option(FailsafeOption opt) const
+{
+    return (g2.fs_options & (uint32_t)opt);
+}
+
 void Copter::failsafe_radio_on_event()
 {
-    // if motors are not armed there is nothing to do
-    if( !motors->armed() ) {
-        return;
-    }
+    AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_RADIO, LogErrorCode::FAILSAFE_OCCURRED);
 
+    int16_t action;
+    action = g.failsafe_throttle; //The action set in FS_THR_ENABLE
+
+    // Conditions to deviate from FS_THR_ENABLE selection and sent specigic GCS warning
     if (should_disarm_on_failsafe()) {
+        // should immediately disarm when we're on the ground
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - DISARMING");
         arming.disarm();
+        action = FS_THR_DISABLED;
+
+    } else if (flightmode->is_landing() && ((battery.has_failsafed() && battery.get_highest_failsafe_priority() <= FAILSAFE_LAND_PRIORITY))) {
+        // Allow landing to continue when battery failsafe requires it (not a user option)
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO + BATTERY FAILSAFE - CONTINUING LANDING");
+        action = FS_THR_ENABLED_ALWAYS_LAND;
+
+    } else if (flightmode->is_landing() && failsafe_option(FailsafeOption::THR_LAND_CONTINUE)) {
+        // Allow landing to continue when FS_OPTIONS is set to continue landing
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - CONTINUING LANDING");
+        action = FS_THR_ENABLED_ALWAYS_LAND;
```

Comment:
In general I think we should consolidate on "Radio Failsafe" instead of "RC" or "Throttle" failsafe.  This enum may be pre-existing though so this isn't a blocker.

Reply:
I agree.  The terms radio, rc, and throttle are used interchangeably throughout the code and documentation. That is an existing enum. I would certainly be in favor of making it all consistent.  I would actually favor RC over radio since RC is more specific than radio. Radio could be too open ended... which radio, i have like 4 on my copter :).

This would mean changing EVERYTHING.  Even the existing parameter from `FS_THR_ENABLE` to something like `FS_RC_FAILSAFE`.  I can do this, as long as that amount of changing things is desired.

Reply:
`FS_THR_...` is used about 100 times throughout the code. So I'm not this would be a good idea to change to something else.